### PR TITLE
Fix generic instance for nullary constructors

### PR DESCRIPTION
**Pull request by cdparks**
**tldr;** `ghc` couldn't find the nil `GHasDependenciesRecursive` instance for nullary constructors because of some hidden ambiguity.

#### Slightly Longer
Before, we had this weird list type:
```haskell
data List a
  = None
  | Last a
  | Cons a (List a)
```

You'll note that `None` doesn't bind `a`, so if you have an empty list, the type of `'None` is polymorphic. One way to fix this is to change `None` to `None a` and have `FindMatches` do this:
```haskell
FindMatches nodeTy depsTy () () = 'None ('NoMatch ())
```

Then, you can write this instance for `GHasDependenciesRecursive` which `ghc` can suddenly find:
```haskell
instance GHasDependenciesRecursive (Proxy ('None ('NoMatch ()))) () () where
  genericDependsOnRecursive _ () () = ()
```

But that's kind of gross.

Instead, adding a kind signature to both `FindMatches` and the null instance for `GHasDependenciesRecursive` resolves whatever ambiguity was tripping up `ghc`. In fact, we can now use the builtin type-level list. I suspect this same ambiguity is what made me add the `Last` constructor to `List` in the first place.

### CODE REVIEW

**Comment by eborden** [comment]
¡Que bueno!

Great job. I'm happy to see more built-ins and more red than green!

**Comment by eborden** [comment]
Thanks @cbeav for finding this.

**Comment by cbeav** [comment]
Thanks @eborden for telling me I'd found this :)

**Comment by cdparks** [comment]
Tested on our API tests, which still work.
